go/storage/mkvs/checkpoint/chunk.go

```diff
@@ -16,65 +18,213 @@ import (
+
+		chunks = append(chunks, chunkHash)
+
+		// Check if we are finished.
+		if nextOffset == nil {
+			break
+		}
+	}
+
+	return chunks, nil
+}
+
+func (sc *seqChunker) createChunk(ctx context.Context, tree mkvs.Tree, offset node.Key, w io.Writer) (
 	chunkHash hash.Hash,
 	nextOffset node.Key,
 	err error,
 ) {
 	it := tree.NewIterator(
 		ctx,
 		// V1 checkpoints use V0 proofs.
-		mkvs.WithProofBuilder(syncer.NewProofBuilderV0(root.Hash, root.Hash)),
+		mkvs.WithProofBuilder(syncer.NewProofBuilderV0(sc.root.Hash, sc.root.Hash)),
 	)
 	defer it.Close()
 
 	// We build the chunk until the proof becomes too large or we have reached the end.
-	for it.Seek(offset); it.Valid() && it.GetProofBuilder().Size() < chunkSize; it.Next() {
+	for it.Seek(offset); it.Valid() && it.GetProofBuilder().Size() < sc.chunkSize; it.Next() {
 		// Check if context got cancelled while iterating to abort early.
 		if ctx.Err() != nil {
 			err = ctx.Err()
 			return
 		}
 	}
 	if it.Err() != nil {
-		err = fmt.Errorf("chunk: failed to iterate: %w", it.Err())
+		err = fmt.Errorf("failed to iterate: %w", it.Err())
 		return
 	}
 
 	// Build our chunk.
 	proof, err := it.GetProof()
 	if err != nil {
-		err = fmt.Errorf("chunk: failed to build proof: %w", err)
+		err = fmt.Errorf("failed to build proof: %w", err)
 		return
 	}
 
 	// Determine the next offset (not included in proof).
-	it.Next()
-	nextOffset = it.Key()
+	if it.Valid() {
+		it.Next()
+		nextOffset = it.Key()
+	}
+
+	chunkHash, err = writeChunk(proof, w)
+	return
+}
+
+type parallChunker struct {
+	ndb       db.NodeDB
+	root      node.Root
+	chunkSize uint64
+	threads   uint16
+}
+
+func (pc *parallChunker) chunk(ctx context.Context, wf writerFactory) ([]hash.Hash, error) {
+	root, err := newSubtree(pc.ndb, pc.root)
+	if err != nil {
+		return nil, err
+	}
+	pending := []*subtree{root}
+
+	// chunking must be deterministic!
+	var chunks []hash.Hash
+	for len(pending) > 0 {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		pending, err = pc.splitTasks(pending, 0)
+		if err != nil {
+			return nil, fmt.Errorf("chunk: splitting chunking tasks: %w", err)
+		}
+		fmt.Println(len(pending))
+
+		hashes, err := pc.createChunks(ctx, wf, pending)
+		if err != nil {
+			return nil, fmt.Errorf("chunk: processing chunking tasks: %w", err)
+		}
+
+		chunks = append(chunks, hashes...)
 
+		pending = pc.filterFinished(pending)
+	}
+
+	return chunks, nil
+}
+
+func (pc *parallChunker) splitTasks(tasks []*subtree, tried int) ([]*subtree, error) {
+	if tried > 10 {
+		return tasks, nil
+	}
+
+	var tasksN []*subtree
+	for i, task := range tasks {
+		if len(tasksN)+len(tasks)-i >= int(pc.threads) {
+			tasksN = append(tasksN, tasks[i:]...)
+			break
+		}
+		children, err := task.split()
+		if err != nil {
+			return nil, fmt.Errorf("splitting chunking task: %w", err)
+		}
+		tasksN = append(tasksN, children...)
+	}
+
+	if len(tasksN) < int(pc.threads) {
+		return pc.splitTasks(tasksN, tried+1)
+	}
```

Comment:
Why recursion? Given that it is only at the end it should be easy to refactor to an iterative version.